dali/kernels/reduce/reduce_gpu_impl.cuh

```diff
@@ -1095,72 +1101,289 @@ class ReduceImplGPU {
+         return cpu_samples[i].n_reduced == 1;
+        },
+        [&](int i) {
+         return cpu_samples[i].n_reduced < 1024 && cpu_samples[i].num_macroblocks == 1;
+        },
+        [&](int i) {
+          return cpu_samples[i].n_inner < 32;
+        });
+    auto none_end = std::get<0>(groups);
+    auto middle_small_end = std::get<1>(groups);
+    auto middle_large_inner_small_end = std::get<2>(groups);
+
+    int num_none = none_end - indices.begin();
+    int num_middle_small = middle_small_end - none_end;
+    int num_middle_large_inner_small = middle_large_inner_small_end - middle_small_end;
+    int num_middle_large_inner_medium = indices.end() - middle_large_inner_small_end;
+
     auto *pre = GetPreprocessorBanks<is_first, 2>(wa, stage.axis);
     auto *post = GetPostprocessors<is_last>(wa);
 
-    using pre_bank_t = std::remove_cv_t<std::remove_reference_t<decltype(*pre)>>;;
-    using post_t = std::remove_cv_t<std::remove_reference_t<decltype(*post)>>;;
-    using red_t = std::remove_reference_t<decltype(This().GetReduction())>;
-
-    int max_block_size = std::min(1024, MaxThreadsPerBlock(
-      ReduceMiddleKernel<Acc, StageOut, StageIn, red_t, pre_bank_t, post_t>));
+    permute_in_place(cpu_samples, indices);
+    if (pre)
+      permute_in_place(pre, indices);
+    if (post)
+      permute_in_place(post, indices);
 
     wa.CopyParamsToDevice(ctx.stream);
-    dim3 block(32, max_block_size / 32);
-    int gridx = std::max(32, 512/stage.num_samples());
-    dim3 grid(gridx, stage.num_samples());
-    const int shm_size = 0x8000;  // 32 kB shared mem
 
     SampleDesc *gpu_samples = wa.GetDeviceParam(cpu_samples);
     auto *gpu_pre = wa.GetDeviceParam(pre);
     auto *gpu_post = wa.GetDeviceParam(post);
 
-    ReduceMiddleKernel<Acc><<<grid, block, shm_size, ctx.stream>>>(
-      gpu_samples, This().GetReduction(), gpu_pre, gpu_post);
+    using pre_bank_t = std::remove_cv_t<std::remove_reference_t<decltype(*pre)>>;;
+    using post_t = std::remove_cv_t<std::remove_reference_t<decltype(*post)>>;;
+    using red_t = std::remove_reference_t<decltype(This().GetReduction())>;
+
+    auto launch_params = [&](auto kernel, int nsamples, int shm_size, int max_block_size) {
+      int preferred_block_size = max_block_size;
+      int preferred_grid_size;  // unused
+      CUDA_CALL(cudaOccupancyMaxPotentialBlockSize(
+        &preferred_grid_size,
+        &preferred_block_size,
+        kernel,
+        shm_size,
+        max_block_size));
+
+      dim3 block(32, preferred_block_size / 32);
+      int gridx = std::max(32, 512/nsamples);
+      dim3 grid(gridx, nsamples);
+      return std::make_pair(grid, block);
+    };
+
+    dim3 grid, block;
+    int sample_offset = 0;
+
+    // None
+    if (num_none) {
+      std::tie(grid, block) = launch_params(
+          ReduceNoneKernel<Acc, StageOut, StageIn, red_t, pre_bank_t, post_t>,
+          num_none, 0, 256);
+
+      ReduceNoneKernel<Acc><<<grid, block, 0, ctx.stream>>>(
```

Comment:
I know this is repetitive. I've tried hoisting the parameter setup _and_ launch to a function, but CUDA kernels aren't first-class C++ entities and cannot be passed as a template argument - at that point they become ordinary functions, so I'd have to launch them with `cudaKernelLaunch` instead of `<<<>>>`, losing compile-time parameter checking.